implementations/rust/ockam/ockam_command/src/enroll/auth0_service.rs

```diff
@@ -0,0 +1,526 @@
+                    if let Some(spinner) = spinner_option.as_ref() {
+                        spinner.finish_and_clear();
+                    }
+                    opts.terminal.write_line(&fmt_para!("Authenticated\n"))?;
+                    return Ok(token);
+                }
+                _ => {
+                    let err = res.json::<TokensError>().await.into_diagnostic()?;
+                    match err.error.borrow() {
+                        "authorization_pending" | "invalid_request" | "slow_down" => {
+                            debug!(?err, "tokens not yet received");
+                            sleep(Duration::from_secs(dc.interval as u64)).await;
+                            continue;
+                        }
+                        _ => {
+                            let err_msg = "failed to receive tokens";
+                            debug!(?err, "{err_msg}");
+                            return Err(miette!(err_msg).into());
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Generate 32 random bytes as a code verifier
+    // to prove that this client was really the one requesting an authorization code
+    /// See the full PKCE flow here: https://datatracker.ietf.org/doc/html/rfc7636
+    fn create_code_verifier(&self) -> String {
+        let mut code_verifier = [0u8; 32];
+        let mut rng = thread_rng();
+        rng.fill_bytes(&mut code_verifier);
+        base64_url::encode(&code_verifier)
+    }
+
+    /// Return the list of scopes for the authorization requests
+    fn scopes(&self) -> String {
+        "profile openid email".to_string()
+    }
+
+    /// Extract the `code` query parameter from the callback request
+    fn get_code(request_url: &str) -> Result<String> {
+        // The local url is retrieved as a path associated to the tiny-http request
+        // In order to parse it with the Url parser we need to first recover a full URL
+        let url =
+            Url::parse(format!("http://0.0.0.0:0{}", request_url).as_str()).into_diagnostic()?;
+
+        // Check the URL path
+        if !url.path().starts_with("/callback") {
+            return Err(miette!(
+                "the query path should be of the form '/callback?code=xxxx'. Got: {})",
+                request_url
+            )
+            .into());
+        };
+
+        // Extract the 'code' query parameter
+        if let Some((name, value)) = url.query_pairs().next() {
+            if name == "code" {
+                return Ok(value.to_string());
+            };
+        };
+        Err(miette!(
+            "could not extract the 'code' query parameter from the path {})",
+            request_url
+        )
+        .into())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// This test can only run with an open browser in order to authenticate the user
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    #[ignore]
```

Comment:
`ignore` can optionally take a string description - we could use that to summarize why a given test is opt-in, like we currently used comments for:
https://doc.rust-lang.org/reference/attributes/testing.html#the-ignore-attribute

Reply:
Thanks, done.